src/uu/env/src/raw_string_parser.rs

```diff
@@ -0,0 +1,307 @@
+// This file is part of the uutils coreutils package.
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+//
+//! SAFETY: This module does "unsafe" byte by byte operations on a UTF8 encoded string.
+//! UTF8 encodes all non-ASCII characters as multi-byte characters. Meaning that the UTF8
+//! string contains short sequences of bytes which should not be splitted or individually modified.
+//! All bytes that belong to a multi-byte character sequence are defined to have a different value
+//! than any ASCII single byte char. This can be used to easily detect where multi-byte character sequences
+//! start and end.
+//! To guarantee that after processing the output is again valid UTF8, the following rules must apply:
+//! 1. Move multi-byte characters as a whole.
+//! 2. Insert characters only on ASCII boundaries.
+//! We also want to support even strings that contain partially invalid utf8. Thats why we can't rely
+//! on std library functionality when dealing with multi-byte characters.
+//!
+//! The general idea of this module is to encapsulate the unsafe parts in a small and easily testable unit.
+// spell-checker:ignore (words) splitted
+#![allow(unsafe_code)]
+
+use std::mem;
+
+pub fn is_ascii(c: u8) -> bool {
+    (c & 0x80) == 0
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Error {
+    pub look_at_pos: usize,
+    pub err_type: ErrorType,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ErrorType {
+    NoAsciiBoundary,
+    NoAsciiChar,
+    NoAsciiCharInput,
+    EndOfInput,
+    InternalError,
+}
+
+pub struct RawStringParser<'a> {
+    pub input: &'a str,
+    pointer: usize,
+    pointer_str: &'a str, // just for debugging sessions. In release build it will be removed by the compiler.
```

Comment:
How do we know that it will be removed? I'd rather not have it to be honest.

Reply:
My experience with compilers in C/C++ tells me that if something is not used to generate output, it's typically removed. Compilers are quite powerfull/smart in optimizing.

But current state of the code is that its actually used. So its not so sure anymore if the compiler is removing it. I just removed the comment. Hope this is fine then.